src/mp/sys/oid.c

```diff
@@ -822,40 +824,69 @@ MpOidWatchdogIsExpired(
 _IRQL_requires_max_(PASSIVE_LEVEL)
 NTSTATUS
 MpIrpOidSetFilter(
-    _In_ ADAPTER_CONTEXT *Adapter,
+    _In_ EXCLUSIVE_USER_CONTEXT *UserContext,
     _In_ IRP *Irp,
     _In_ IO_STACK_LOCATION *IrpSp
     )
 {
     NTSTATUS Status;
-    OID_KEY *Keys;
+    ADAPTER_CONTEXT *Adapter = UserContext->Adapter;
+    UINT32 InKeySize;
+    const VOID *InKeys;
+    OID_KEY *OidFilterKeys = NULL;
     UINT32 KeyCount;
     KIRQL OldIrql = PASSIVE_LEVEL;
     BOOLEAN IsLockHeld = FALSE;
 
-    if (IrpSp->Parameters.DeviceIoControl.InputBufferLength < sizeof(*Adapter->OidFilterKeys)) {
+    if (UserContext->Header.ApiVersion >= 2) {
+        InKeySize = sizeof(OID_KEY);
+    } else {
+        InKeySize = sizeof(OID_KEY_V0);
+    }
+
+    if (IrpSp->Parameters.DeviceIoControl.InputBufferLength < InKeySize) {
         Status = STATUS_INVALID_PARAMETER;
         goto Exit;
     }
 
-    if (IrpSp->Parameters.DeviceIoControl.InputBufferLength % sizeof(*Adapter->OidFilterKeys) != 0) {
+    if (IrpSp->Parameters.DeviceIoControl.InputBufferLength % InKeySize != 0) {
         Status = STATUS_INVALID_PARAMETER;
         goto Exit;
     }
 
-    Keys = (OID_KEY *)Irp->AssociatedIrp.SystemBuffer;
-    KeyCount = IrpSp->Parameters.DeviceIoControl.InputBufferLength / sizeof(*Adapter->OidFilterKeys);
+    InKeys = Irp->AssociatedIrp.SystemBuffer;
+    KeyCount = IrpSp->Parameters.DeviceIoControl.InputBufferLength / InKeySize;
+
+    OidFilterKeys =
+        ExAllocatePoolZero(
+            NonPagedPoolNx, KeyCount * sizeof(*Adapter->OidFilterKeys), POOLTAG_MP_OID);
+    if (OidFilterKeys == NULL) {
+        Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
 
     for (UINT32 Index = 0; Index < KeyCount; Index++) {
-        OID_KEY *Key = &Keys[Index];
-        if (Key->RequestType != NdisRequestQueryInformation &&
-            Key->RequestType != NdisRequestSetInformation &&
-            Key->RequestType != NdisRequestMethod) {
+        const VOID *InAnyKey = RTL_PTR_ADD(InKeys, (UINT32)(Index * InKeySize));
+        OID_KEY *FilterKey = &OidFilterKeys[Index];
+
+        if (UserContext->Header.ApiVersion >= 2) {
```

Comment:
Should these current/latest supported API version checks use ```FNLWF_IOCTL_CURRENT_VERSION``` instead of the hardcoded literal for increased maintainability?

Reply:
I was thinking about the best way to do this, and FNLWF_IOCTL_CURRENT_VERSION doesn't work because it keeps changing over time. We could `#define` specific macros for each of these iterations, which might be at least more readable, even if the behavior is the same at the end of the day.

From a maintenance point of view, I guess the most important thing would be to be able to find and replace all version checks less than specific number, if we release a new major version that breaks back compat.

Reply:
Ah for some reason I was only thinking about the conditional update, and not both the conditional and the body update. Yeah after thinking about it ```FNLWF_IOCTL_CURRENT_VERSION``` doesn't solve it

Reply:
I added macros so we can do `LWF_APIVER(2)` instead of `2` which makes our code quite a bit more maintainable IMO, without requiring any extra boilerplate to add a new version. I strongly dislike how some components (e.g. NDIS) are full of hundreds or thousands of defines like `#define NDIS_VERSION_600 600` and `#define NDIS_STRUCT_VER_1 1` and so on, and want to avoid that in such a lightweight project like this.